dropshot/examples/file_server.rs

```diff
@@ -0,0 +1,213 @@
+    rqctx: Arc<RequestContext<FileServerContext>>,
+    path: Path<AllPath>,
+) -> Result<Response<Body>, HttpError> {
+    let path = path.into_inner().path;
+    let mut entry = rqctx.context().base.clone();
+    for component in &path {
+        /* The previous iteration needs to have resulted in a directory. */
+        if !entry.is_dir() {
+            return Err(HttpError::for_bad_request(
+                None,
+                "ENOTDIR".to_string(),
+            ));
+        }
+        /* Dropshot won't ever give us dot-components. */
+        assert_ne!(component, ".");
+        assert_ne!(component, "..");
+        entry.push(component);
+
+        /*
+         * We explicitly prohibit consumers from following symlinks to prevent
+         * showing data outside of the intended directory.
+         */
+        let m = entry.symlink_metadata().map_err(|_| {
+            HttpError::for_bad_request(None, "ENOENT".to_string())
+        })?;
+        if m.file_type().is_symlink() {
+            return Err(HttpError::for_bad_request(None, "EMLINK".to_string()));
+        }
+    }
+
+    /*
+     * If the entry is a directory, we serve a listing of its contents. If it's
+     * regular file we serve the file.
+     */
+    if entry.is_dir() {
+        let body = dir_body(entry).await.map_err(|_| {
+            HttpError::for_bad_request(None, "EBADF".to_string())
+        })?;
+
+        Ok(Response::builder()
+            .status(StatusCode::OK)
+            .header(http::header::CONTENT_TYPE, "text/html")
+            .body(body.into())?)
+    } else {
+        let body = tokio::fs::read(&entry).await.map_err(|_| {
+            HttpError::for_bad_request(None, "EBADF".to_string())
+        })?;
+
+        /* Derive the MIME type from the file name */
+        let content_type = mime_guess::from_path(&entry)
+            .first()
+            .map_or_else(|| "text/plain".to_string(), |m| m.to_string());
+
+        Ok(Response::builder()
+            .status(StatusCode::OK)
+            .header(http::header::CONTENT_TYPE, content_type)
+            .body(body.into())?)
+    }
+}
+
+/**
+ * Generate a simple HTML listing of files within the directory.
+ * See the note below regarding the handling of trailing slashes.
+ */
+async fn dir_body(dir_path: PathBuf) -> Result<String, std::io::Error> {
+    let dir_link = dir_path.to_string_lossy();
+    let mut dir = tokio::fs::read_dir(&dir_path).await?;
+
+    let mut body = String::new();
+
+    body.push_str(
+        format!(
+            "<html>
+            <head><title>{}/</title></head>
+            <body>
+            <h1>{}/</h1>
+            ",
+            dir_link, dir_link
+        )
+        .as_str(),
+    );
+    body.push_str("<ul>\n");
+    while let Some(entry) = dir.next_entry().await? {
+        let name = entry.file_name();
+        let name = name.to_string_lossy();
+        /*
+         * Note that Dropshot handles paths with and without trailing slashes
```

Comment:
(Meta: I haven't used github for code review in a serious way in my life, so let me know if I'm holding any of the comment/review features strangely please!)

Yeah, per my comment above, I think the details of path handling are maybe the most interesting thing to see in this PR, besides the fact of just being an example. Things like /./ as a middle component too, though I see dropshot handles that.

Reply:
Do you mean in terms of a new test for the `input_path_to_segments` function?

Reply:
I think based on my understanding now, that's just well understood to be the behavior of dropshot, so I don't think you need to add any particular testing of it in a thing using dropshot.

Reply:
I think based on my understanding now, that's just well understood to be the behavior of dropshot, so I don't think you need to add any particular testing of it in a thing using dropshot.